### PR TITLE
10145 upgrade comment type/attr.ib annotations to native type annotations 

### PR DESCRIPTION
**Pull request by graingert**
## Scope and purpose

Python 3.6 and attr.s both support native variable and field annotation, so use them

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10145
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [ ] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [ ] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] I have added `twisted/twisted-contributors` teams to the PR `Reviewers`.
* [ ] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: <github_username>, <github_usernames_if_more_authors>
Reviewer: <github_username>, <github_usernames_if_more_reviewers>
Fixes: ticket:<trac_ticket_number>, ticket:<another_if_more_in_one_PR>

Long description providing a summary of these changes.
(as long as you wish)
```

### CODE REVIEW

File: src/twisted/internet/base.py [hunk 1]
```diff
@@ -335,7 +335,7 @@ def getHostByName(
             timeoutDelay = sum(timeout)
         else:
             timeoutDelay = 60
-        userDeferred = defer.Deferred()  # type: Deferred[str]
+        userDeferred: Deferred[str] = defer.Deferred()
```
Comment (graingert) [inline src/twisted/internet/base.py]: I think this form is preferred:

```suggestion
        userDeferred = defer.Deferred[str]()
```

Comment (wsanchez) [reply src/twisted/internet/base.py]: Even better, get rid of the `defer.` prefix

File: src/twisted/internet/address.py [hunk 2]
```diff
@@ -35,13 +36,15 @@ class IPv4Address:
     @type port: C{int}
     """
 
-    type = attr.ib(validator=attr.validators.in_(["TCP", "UDP"]))
-    host = attr.ib()
-    port = attr.ib()
+    type: Union[Literal["TCP"], Literal["UDP"]] = attr.ib(
```
Comment (wsanchez) [inline src/twisted/internet/address.py:39]: nice

File: src/twisted/internet/tcp.py [hunk 3]
```diff
@@ -934,8 +935,13 @@ def __exit__(excType, excValue, traceback):
         """
 
 
+class _HasClose(typing_extensions.Protocol):
+    def close(self) -> object:
```
Comment (wsanchez) [inline src/twisted/internet/tcp.py:939]: Not `-> None`?

**Review by wsanchez — APPROVED** [review]

File: src/twisted/internet/tcp.py (hunk 3 above)
Comment (graingert) [reply src/twisted/internet/tcp.py:939]: `-> object` is less restrictive, the consumer doesn't care about the return value at all

**Pull request merged**